crates/shared/src/price_estimation/buffered.rs

```diff
@@ -0,0 +1,450 @@
+/// native prices requests.
+#[derive(Clone)]
+pub struct BufferedTransport<Inner> {
+    config: Configuration,
+    #[allow(dead_code)]
+    inner: Arc<Inner>,
+    calls: mpsc::UnboundedSender<H160>,
+    broadcast_sender: broadcast::Sender<NativePriceResult>,
+}
+
+type NativePriceResult = (H160, Result<f64, PriceEstimationError>);
+
+impl<Inner> BufferedTransport<Inner>
+where
+    Inner: NativePriceBatchFetcher + Send + Sync + 'static,
+{
+    /// Maximum capacity of the broadcast channel, the messages are discarded as
+    /// soon as they are sent, so this limit should be enough to hold the
+    /// flow
+    const BROADCAST_CHANNEL_CAPACITY: usize = 50;
+
+    /// Creates a new buffered transport with the specified configuration.
+    pub fn with_config(inner: Inner, config: Configuration) -> Self {
+        let inner = Arc::new(inner);
+        let (calls, receiver) = mpsc::unbounded();
+
+        let (broadcast_sender, _) = broadcast::channel(Self::BROADCAST_CHANNEL_CAPACITY);
+
+        Self::background_worker(
+            inner.clone(),
+            config.clone(),
+            receiver,
+            broadcast_sender.clone(),
+        );
+
+        Self {
+            inner,
+            calls,
+            broadcast_sender,
+            config,
+        }
+    }
+
+    /// Start a background worker for handling batched requests.
+    fn background_worker(
+        inner: Arc<Inner>,
+        config: Configuration,
+        calls: mpsc::UnboundedReceiver<H160>,
+        broadcast_sender: broadcast::Sender<NativePriceResult>,
+    ) -> JoinHandle<()> {
+        tokio::task::spawn(batched_for_each(config, calls, move |batch| {
+            let inner = inner.clone();
+            let broadcast_sender = broadcast_sender.clone();
+            async move {
+                let batch = batch.into_iter().collect::<HashSet<_>>();
+                if !batch.is_empty() {
+                    let results = match inner.fetch_native_prices(&batch).await {
+                        Ok(results) => results
+                            .into_iter()
+                            .map(|(token, price)| (token, Ok::<_, PriceEstimationError>(price)))
+                            .collect::<HashMap<_, _>>(),
+                        Err(err) => {
+                            tracing::error!(?err, "failed to send native price batch request");
+                            batch
+                                .into_iter()
+                                .map(|token| (token, Err(err.clone())))
+                                .collect::<HashMap<_, _>>()
+                        }
+                    };
+                    for result in results {
+                        let _ = broadcast_sender.send(result);
+                    }
+                }
+            }
+        }))
+    }
+
+    /// Blocking operation to get estimate prices in a batch
+    pub async fn blocking_buffered_estimate_prices(
```

Comment:
Which specific operation is considered blocking here? Since Tokio uses asynchronous operations (async/await), it should not block the thread, allowing the scheduler to utilize the thread for other tasks while awaiting. Could you clarify this?

Reply:
@squadgazzz that is a really good question. It is not blocking in the sense of thread blocking. I called it like that because it is a "fire and wait" type of request. But you are right, the name is quite confusing, I'll change it! :ok_hand: